Library/Homebrew/livecheck/strategy/pypi.rb

```diff
@@ -19,8 +19,21 @@ module Strategy
       class Pypi
         NICE_NAME = "PyPI"
 
+        # The `Regexp` used to extract the package name and suffix (e.g., file
+        # extension) from the URL basename.
+        FILENAME_REGEX = /
+          (?<package_name>.+)- # The package name followed by a hyphen
+          .*? # The version string
+          (?<suffix>\.tar\.[a-z0-9]+|\.[a-z0-9]+)$ # Filename extension
+        /ix.freeze
+
         # The `Regexp` used to determine if the strategy applies to the URL.
-        URL_MATCH_REGEX = /files\.pythonhosted\.org/i.freeze
+        URL_MATCH_REGEX = %r{
+          ^https?://files\.pythonhosted\.org
+          /packages
+          (?:/[^/]+)+ # The hexadecimal paths before the filename
```

Comment:
Exactly three of them so maybe it is a good idea to be strict about it:
```suggestion
          (?:/[^/]+){3} # The hexadecimal paths before the filename
```

Reply:
I'm not sure that there's a clear benefit to being explicit about this. We're not interested in the hexadecimal paths and it doesn't matter all that much how many there are. If it suddenly changed to any number other than three in the future (unlikely but possible), this would quietly fail to match those URLs until we manually update it.

This is just my own view but it seems like it would be better not to make more work for ourselves if we don't stand to benefit in some way from this being more explicit. We usually try to strike a balance between being too loose and too explicit, so I think it may be better to wait until this particular looseness creates a demonstrable problem for us.

Reply:
> it doesn't matter all that much how many there are. If it suddenly changed to any number other than three in the future (unlikely but possible), this would quietly fail to match those URLs until we manually update it.
> 
> This is just my own view but it seems like it would be better not to make more work for ourselves if we don't stand to benefit in some way from this being more explicit.

What you just described I see as an important communication signal. It would be good to be explicitly notified once PyPI decides to change assumptions we are building on (which I do find highly unlikely and almost impossible - it would require pip refactoring) because they might be built-in elsewhere in the codebase too. So for me, the _benefit_ is exactly in detecting when things start silently rearranging under our feet :upside_down_face: 
Yes, you can say it doesn't matter because if you make it less strict no action is needed but that is a risky take. Sometimes less strict might also mean with potential _bug inside_ and figuring it out might turn into a nightmare because you wouldn't be aware that outside foundations changed in the first place. I guess it could be considered defensive coding and FWIW compared to laziness and adaptability it counts more on my personal scale. :upside_down_face:

Reply:
I understand your view but I don't think the additional strictness is beneficial from the standpoint of livecheck. If other parts of `brew` make assumptions that are no longer true after a URL format change, then they're likely to loudly fail anyway. We shouldn't be making strategies more fragile for the sake of this kind of alternative purpose.

All that we're concerned about in this context is whether `#find_versions` successfully finds versions and the data is appropriate. If the URL format changes and the looser regex allows this to continue working without modifications, then we've saved ourselves some work. This would mean that the URL changes were in a part that doesn't impact the functionality of `#find_versions`. If the URL format changes and the strategy fails, then we'll know about it anyway (i.e., we don't need to make the regex more explicit to achieve this).

The goal of `URL_MATCH_REGEX` is to ensure that the URL being provided to `#find_versions` will work in that method. However, the regex is only as explicit as it needs to be to achieve this without introducing false positives. The number of hexadecimal directories doesn't impact the behavior of `Pypi#find_versions`, so we don't need to formalize it at this point.

If the looseness in this part of the regex ends up impacting the behavior of `#find_versions` in the future, then we'll see this issue in livecheck's output and we simply can address it in the strategy.

Reply:
> We shouldn't be making strategies more fragile for the sake of this kind of alternative purpose.

Agree.

>  This would mean that the URL changes were in a part that doesn't impact the functionality of `#find_versions`. If the URL format changes and the strategy fails, then we'll know about it anyway (i.e., we don't need to make the regex more explicit to achieve this).

Hmm, the latter is definitely true but I'm not universally sure about the former. I think it is still possible to make changes to the URL in a part that does in fact impact `#find_versions` but slips through `#match?`.

1. Now I need to prove it which requires coming up with some crazy clever pattern just for the sake of breaking it but I'm lazy :grin:  and 
2. We already agreed it doesn't matter at all because it would fail loudly anyway

Bottom line, from a practical standpoint you are right, and thank you for patiently addressing my philosophical arguments.